aptos-move/framework/aptos-framework/sources/permissioned_signer.move

```diff
@@ -0,0 +1,620 @@
+                    true
+                } else { false }
+            }
+            StoredPermission::Unlimited => true
+        }
+    }
+
+    /// increase `threshold` capacity from StoredPermission
+    fun increase_capacity(perm: &mut StoredPermission, threshold: u256) {
+        match (perm) {
+            StoredPermission::Capacity(current_capacity) => {
+                *current_capacity = *current_capacity + threshold;
+            }
+            StoredPermission::Unlimited => (),
+        }
+    }
+
+    /// merge the two stored permission
+    fun merge(lhs: &mut StoredPermission, rhs: StoredPermission) {
+        match (rhs) {
+            StoredPermission::Capacity(new_capacity) => {
+                match (lhs) {
+                    StoredPermission::Capacity(current_capacity) => {
+                        *current_capacity = *current_capacity + new_capacity;
+                    }
+                    StoredPermission::Unlimited => (),
+                }
+            }
+            StoredPermission::Unlimited => *lhs = StoredPermission::Unlimited,
+        }
+    }
+
+    /// =====================================================================================================
+    /// Permission Management
+    ///
+    /// Authorizes `permissioned` with the given permission. This requires to have access to the `master`
+    /// signer.
+
+    inline fun map_or<PermKey: copy + drop + store, T>(
+        permissioned: &signer,
+        perm: PermKey,
+        mutate: |&mut StoredPermission| T,
+        default: T,
+    ): T {
+        let permission_signer_addr = permission_address(permissioned);
+        assert!(
+            exists<PermissionStorage>(permission_signer_addr),
+            error::permission_denied(E_NOT_ACTIVE)
+        );
+        let perms =
+            &mut borrow_global_mut<PermissionStorage>(permission_signer_addr).perms;
+        let key = copyable_any::pack(perm);
+        if (simple_map::contains_key(perms, &key)) {
+            mutate(simple_map::borrow_mut(perms, &key))
+        } else {
+            default
+        }
+    }
+
+    inline fun insert_or<PermKey: copy + drop + store>(
+        permissioned: &signer,
+        perm: PermKey,
+        mutate: |&mut StoredPermission|,
+        default: StoredPermission,
+    ) {
+        let permission_signer_addr = permission_address(permissioned);
+        assert!(
+            exists<PermissionStorage>(permission_signer_addr),
+            error::permission_denied(E_NOT_ACTIVE)
+        );
+        let perms =
+            &mut borrow_global_mut<PermissionStorage>(permission_signer_addr).perms;
+        let key = copyable_any::pack(perm);
+        if (simple_map::contains_key(perms, &key)) {
+            mutate(simple_map::borrow_mut(perms, &key));
+        } else {
+            simple_map::add(perms, key, default);
+        }
+    }
+
+    /// Authorizes `permissioned` with a given capacity and increment the existing capacity if present.
```

Comment:
General remark: it is kind of hard to read this code as a public API because public functions are hidden in a lot of private and friend code. Is it possible to move the public functions somewhere in front? I think we should expect that users of this module will really look at this code here for a high-level understanding. Consider that main 'user' of the code is not someone who changes it, but someone who calls it.

Reply:
I agree, I think we do have a convention in all of framework code to put private and public(friend) functions below the fold, and put public functions at the time.